### PR TITLE
DOCS: 4.x Pico doc introductions

### PR DESCRIPTION
**Pull request by ljamen**
WIP - need to create PR to stash changes for now.

### CODE REVIEW

**Comment by ljamen** [comment]
@trentjeff Hi Jeff - leaving this as a draft PR until we have a chance to go through the basics and then we need to rename the Nima directory to SE (and remove the existing SE directory). Holding off on doing that while we are working on the other docs.

File: docs/nima/builder.adoc [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+///////////////////////////////////////////////////////////////////////////////
+
+= Helidon Builder
+:description: about Helidon Builder
+:keywords: helidon, java, microservices, reactive, virtual threads, config, builder-config
```
Comment (trentjeff) [inline docs/nima/builder.adoc:21]: remove reactive and add "fluent builder"

File: docs/nima/builder.adoc [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+- <<Usage, Usage>>
+- <<Configuration, Configuration>>
+- <<Examples, Examples>>
+
+//the purpose of this article is to introduce Helidon Builder and provide some high-level information about how and why its used. The examples and eventual Builder guide will provide the detailed steps. 
+
+
+== Overview
+
+Helidon Builder is an annotation processor that code-generates fluent builder support just by adding `@Builder` to your declared Java types. This declarative style of using annotations helps you by reducing the amount of boilerplate [redundant?] code you need to write.
```
Comment (trentjeff) [inline docs/nima/builder.adoc:39]: Helidon builder is an optional feature and offers an annotation processor ...

Add the optional feature bit.

it is not redundant btw.

File: docs/nima/builder.adoc [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+* Interception:  You can provide other custom semantics during the build() method (similar to validation) for other types of interception or decoration on your target builders.
+
+* Packaging and usage choice: Helidon Builder can be used as a standalone module in your build as long as you are using Java 11 or later. When used in a standalone manner in your build, you can use the `requireLibraryDependencies()` attribute to toggle between supporting libraries (like validation). This allows you to either generate code directly into your application, or have it supplied by Helidon libraries. This gives you an opportunity to use Helidon's tooling in isolation in your project build, even if you don't plan to use Helidon in any other part of your application runtime. 
+
+//Do we want to say you don't plan to use Helidon???
```
Comment (trentjeff) [inline docs/nima/builder.adoc:53]: I think its fine the way it is written.

File: docs/nima/builder.adoc [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+To enable Builder, add the following dependencies and annotation processor path to your project's `pom.xml`:
+
+.Builder dependencies
+[source,java]
+
+----
+<dependency>
+   <groupId>io.helidon.builder</groupId>
+   <artifactId>helidon-builder</artifactId>
+   <version>4.0.0-ALPHA4</version>
```
Comment (trentjeff) [inline docs/nima/builder.adoc:66]: ALPHA6 (or just prepare for the M1 version)

File: docs/nima/builder.adoc [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+       <plugin>
+           <groupId>org.apache.maven.plugins</groupId>
+           <artifactId>maven-compiler-plugin</artifactId>
+           <version>3.10.1</version>
+           <configuration>
+               <annotationProcessorPaths>
+                   <path>
+                       <groupId>io.helidon.builder</groupId>
+                       <artifactId>helidon-builder-processor</artifactId>
+                       <version>4.0.0-ALPHA4</version>
```
Comment (trentjeff) [inline docs/nima/builder.adoc:87]: same here

File: docs/nima/builder.adoc [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+----
+
+.Write your main()
+[source, java]
+
+----
+public class Main {
+   public static void main(String[] args) {
+       NameAndDescriptionAndComments target =
+               DefaultNameAndDescriptionAndComments.builder()
```
Comment (trentjeff) [inline docs/nima/builder.adoc:148]: We just reversed the code generation to make "Default" the suffix and not the prefix for generated code.  So technically the name is `NameAndDescriptionAndCommentsDefault`

File: docs/nima/builder.adoc [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+
+=== Auto-Generated Code
+
+The Helidon Builder's annotation processor generates source code whenever it processes an annotation that it recognizes - in this case the @Builder annotation. You must apply the annotation processors in your build in order to generate the supporting sources before they are consumed in the main() from this example.
+
+
+
+
+== Configuration
```
Comment (trentjeff) [inline docs/nima/builder.adoc:176]: ConfigBean Builder

File: docs/nima/builder.adoc [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+//I wasn't sure how to position Builder Config. It seems like it fits in with Builder but perhaps we want to keep it separate? Also, are we using the term Pico Builder Config or something else?
```
Comment (trentjeff) [inline docs/nima/builder.adoc:178]: ConfigBean Builder is an extensibility feature over the base Builder to allow integration of the generated builder types to the Helidon Configuration subsystem.

Comment (trentjeff) [reply docs/nima/builder.adoc:178]: Oh, I see you have it below.

File: docs/nima/builder.adoc [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+Pico Builder Config is a specialization of Builder that extends the builder to support additional integration with Helidon's configuration subsystem. It adds support for the `@ConfigBean` annotation. When applied to a target interface it will map that interface to configuration via a new toBuilder method generated on the implementation as follows:
```
Comment (trentjeff) [inline docs/nima/builder.adoc:180]: Remove Pico

Comment (trentjeff) [reply docs/nima/builder.adoc:180]: also make it `toBuilder()` as a code reference

File: docs/nima/builder.adoc [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+    * `@Builder` can be used in conjunction with `@ConfigBean`. All attributed will be honored exception one...
+    * `Builder.requireLibraryDependencies` is not supported. All generated configuration beans and builders will minimally require a compile-time and runtime dependency on Helidon's common-config module. But for full fidelity support of Helidon's config one should instead use the full config module.
+
+
+== Examples
+
+* See https://github.com/helidon-io/helidon/blob/4.0.0-ALPHA4/builder/tests/builder[tests/builder] for additional Helidon Builder usage examples.
+
+* See https://github.com/helidon-io/helidon/tree/4.0.0-ALPHA4/pico/builder-config/tests[pico/builder-config/tests] for additional usage examples for Builder Configuration.
```
Comment (trentjeff) [inline docs/nima/builder.adoc:206]: remove pico

File: docs/nima/builder.adoc [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+There are a few additional points to understand about ConfigBean and its supporting infrastructure:
+
+    * `@Builder` can be used in conjunction with `@ConfigBean`. All attributed will be honored exception one...
+    * `Builder.requireLibraryDependencies` is not supported. All generated configuration beans and builders will minimally require a compile-time and runtime dependency on Helidon's common-config module. But for full fidelity support of Helidon's config one should instead use the full config module.
+
+
+== Examples
+
+* See https://github.com/helidon-io/helidon/blob/4.0.0-ALPHA4/builder/tests/builder[tests/builder] for additional Helidon Builder usage examples.
```
Comment (trentjeff) [inline docs/nima/builder.adoc:204]: these links should be relative to alpha6

File: docs/nima/pico/pico_intro.adoc [hunk 12]
```diff
@@ -36,11 +36,22 @@
 
 == Overview
 
-Helidon Pico //explain what it is and what it can do. 
+Helidon Pico is an optional feature in Helidon that provides service registry, a lifecycle engine and extensibility for customized code generation. Foundationally, Pico provides a way to develop declarative code using standard Javax/Jakarta annotation types (not Helidon-specific). The Helidon Pico framework provides a mix of declarative and programmatic ways to build your application.
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:39]: provides **a** service registry...

File: docs/nima/pico/pico_intro.adoc [hunk 13]
```diff
@@ -36,11 +36,22 @@
 
 == Overview
 
-Helidon Pico //explain what it is and what it can do. 
+Helidon Pico is an optional feature in Helidon that provides service registry, a lifecycle engine and extensibility for customized code generation. Foundationally, Pico provides a way to develop declarative code using standard Javax/Jakarta annotation types (not Helidon-specific). The Helidon Pico framework provides a mix of declarative and programmatic ways to build your application. 
+
+Pico's minimalist, compile-time generated dependency injection (DI) framework and compile-time source code generation has a number of advantages, including:
+
+* pre-runtime validation of the dependency injection model
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:43]: First bullet should be:
* It enables declarative, Inversion of Control style of programming.

File: docs/nima/pico/pico_intro.adoc [hunk 14]
```diff
@@ -36,11 +36,22 @@
...
+* visibility into your application by providing "less magic" - understandability and debug-ability of your application
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:44]: Offers visibility ..

File: docs/nima/pico/pico_intro.adoc [hunk 15]
```diff
@@ -36,11 +36,22 @@
...
+* deterministic behavior (instead of depending on reflection and classpath ordering, etc.)
+* performance improvements since binding the model at compile-time is more efficient than computing it at runtime.
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:46]: Optimizes performance ...

File: docs/nima/pico/pico_intro.adoc [hunk 16]
```diff
@@ -60,13 +71,76 @@ To enable Pico add the following dependencies to your project's `pom.xml` (see M
...
+    <version>${helidon.version}</version>
+----
+
+.Compile-time dependency:
+[source,java]
+----
+
+  <dependency>
+    <groupId>io.helidon.pico</groupId>
+    <artifactId>helidon-pico-services</artifactId>
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:98]: services -> runtime

File: docs/nima/pico/pico_intro.adoc [hunk 17]
```diff
@@ -60,13 +71,76 @@ To enable Pico add the following dependencies to your project's `pom.xml` (see M
...
+    <version>${helidon.version}</version>
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:99]: this is a better way to refer to versions so you don't need to constantly have APLHA4 -> ALPHA6, M1, etc issues to deal with.

File: docs/nima/pico/pico_intro.adoc [hunk 18]
```diff
@@ -60,13 +71,76 @@ To enable Pico add the following dependencies to your project's `pom.xml` (see M
...
+        MyService myLazyActivatedService = serviceProvider.get();
+
+    * If there are a collection of services requiring activation at startup then we recommend annotating those service implementation types with `RunLevel(RunLevel.STARTUP)` and then use code below in `main()` to lazily activate those services. 
+    
+
+    
+    * While Weight determines list order, the `RunLevel` annotation is used to rank the startup ordering, from the lowest value to the highest value, where `RunLevel.STARTUP == 0`. The developer is expected to activate these directly using code like the following (the get() lazily creates and activates the underlying service type):
+
+      List<ServiceProvider<Object>> startupServices = services
+              .lookup(DefaultServiceInfoCriteria.builder().runLevel(RunLevel.STARTUP).build());
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:127]: Default becomes the suffix.

File: docs/nima/pico/pico_intro.adoc [hunk 19]
```diff
@@ -101,16 +175,26 @@ The Pico API provides types that are generally useful at compile time to assign
 
 == Configuration
 
-Pico Config builder
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:104]: Pico Configuration should be instead referred to as "Pico Config-Driven Services"

File: docs/nima/pico/pico_intro.adoc [hunk 20]
```diff
@@ -101,16 +175,26 @@ The Pico API provides types that are generally useful at compile time to assign
...
+Configuration can be achieved through a Pico add-on that is based on Helidon's configuration subsystem, and adds support for something called _config-driven services_ using the `@ConfiguredBy` annotation.
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:178]: It is not really "Configuration can be achieved".

Config-driven services is more about the second part of the sentence - where the presence of configuration in the Helidon Config subsystem will be injected directly into a service declared to be configured by that configuration type.

File: docs/nima/pico/pico_intro.adoc [hunk 21]
```diff
@@ -101,16 +175,26 @@ The Pico API provides types that are generally useful at compile time to assign
...
 
+When applied to a target service interface, the `@ConfiguredBy` annotation will allow you to use a higher level aggregation for application configuration, and then allow the configuration to drive activation of services in the Pico Framework.
 
+NOTE: You can use Pico configuration in combination with the Helidon configuration subsystem. 
 
-== Examples
+There are a few additional caveats to understand about `ConfiguredBy` and its supporting infrastructure.
 
-//optional 
+* `@ConfigBean Builder` is used to aggregate configuration attributes to this higher-level, application-centric configuration beans.
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:186]: right, this refers us back to the ConfigBean builder section - ideally with a link.

File: docs/nima/pico/pico_intro.adoc [hunk 22]
```diff
@@ -101,16 +175,26 @@ The Pico API provides types that are generally useful at compile time to assign
...
 
-== Additional Information
+* The Pico Framework needs to be started with the supporting configdriven modules in order for configuration to drive service activation.
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:188]: config-driven

File: docs/nima/pico/pico_intro.adoc [hunk 23]
```diff
@@ -101,16 +175,26 @@ The Pico API provides types that are generally useful at compile time to assign
...
 
 
 
 
+== Examples
```
Comment (trentjeff) [inline docs/nima/pico/pico_intro.adoc:193]: When config-driven is used, then a different annotation processor should be used.

The basic pico service APT processor should use this one:

```
<build>
   <plugins>
       <plugin>
           <groupId>org.apache.maven.plugins</groupId>
           <artifactId>maven-compiler-plugin</artifactId>
           <version>3.10.1</version>
           <configuration>
               <annotationProcessorPaths>
                   <path>
                       <groupId>io.helidon.pico</groupId>
                       <artifactId>helidon-pico-processor</artifactId>
                       <version>${helidon.version}</version>
                   </path>
               </annotationProcessorPaths>
           </configuration>
       </plugin>
   </plugins>
</build>
```

whereas the Config-Driven one should use this (this one also covers the basic scenarios as well so you don't need both):

```
<build>
   <plugins>
       <plugin>
           <groupId>org.apache.maven.plugins</groupId>
           <artifactId>maven-compiler-plugin</artifactId>
           <version>3.10.1</version>
           <configuration>
               <annotationProcessorPaths>
                   <path>
                       <groupId>io.helidon.pico.configdriven</groupId>
                       <artifactId>helidon-pico-configdriven-processor</artifactId>
                       <version>${helidon.version}</version>
                   </path>
               </annotationProcessorPaths>
           </configuration>
       </plugin>
   </plugins>
</build>
```

File: docs/nima/builder.adoc (hunk 4 above)
Comment (ljamen) [reply docs/nima/builder.adoc:66]: This should be defined as a variable in the shared dependencies file. I'll make sure we have the correct version attributes set for go-live.

**Pull request closed without merging**